In [11]:
#Q3. Finds longest/shortest sequences and longest ORFs, parsing FASTA by gene identifier (2nd column)

from Bio import SeqIO

START_CODON = "ATG"
STOP_CODONS = {"TAA", "TAG", "TGA"}


def extract_gene_id(full_id):
    parts = full_id.split("|")
    return parts[1] if len(parts) > 1 else full_id

#Find all ORFs in all three forward reading frames.
def find_orfs(seq):
    seq = str(seq).upper()
    frames = {1: [], 2: [], 3: []}

    for frame in range(3):
        start_positions = []
        for i in range(frame, len(seq) - 2, 3):
            codon = seq[i:i+3]
            if codon == START_CODON:
                start_positions.append(i)
            elif codon in STOP_CODONS:
                for start in start_positions:
                    orf_seq = seq[start:i+3]
                    frames[frame+1].append((start, i+3, orf_seq, len(orf_seq)))
                start_positions = []  # reset after stop
    return frames

#Return the longest ORF among all frames.
def find_longest_orf(frames):
    longest = None
    for frame, orfs in frames.items():
        for orf in orfs:
            if longest is None or orf[3] > longest[1][3]:
                longest = (frame, orf)
    return longest

#Main function to analyze FASTA file using gene identifiers.
def analyze_fasta(fasta_file):
    records = list(SeqIO.parse(fasta_file, "fasta"))

    # Build dictionary keyed by gene identifier (2nd column)
    seq_dict = {}
    for rec in records:
        gene_id = extract_gene_id(rec.id)
        seq_dict[gene_id] = str(rec.seq)

    # Determine longest and shortest sequences
    longest_gene = max(seq_dict, key=lambda k: len(seq_dict[k]))
    shortest_gene = min(seq_dict, key=lambda k: len(seq_dict[k]))

    print(f"Longest sequence: {longest_gene} ({len(seq_dict[longest_gene])} bp)")
    print(f"Shortest sequence: {shortest_gene} ({len(seq_dict[shortest_gene])} bp)\n")

    # Find longest ORF per gene
    for gene_id, sequence in seq_dict.items():
        frames = find_orfs(sequence)
        longest_orf = find_longest_orf(frames)
        if longest_orf:
            frame_num, (start, end, orf_seq, length) = longest_orf
            print(f"Gene {gene_id}: Longest ORF in Frame {frame_num}")
            print(f"Start: {start}, End: {end}, Length: {length}")
            print(f"ORF sequence (first 60 bp): {orf_seq[:60]}...\n")
        else:
            print(f"Gene {gene_id}: No valid ORF found\n")


if __name__ == "__main__":
    path = '/content/drive/MyDrive/Biomedical_DS/homework2/'
    fasta_file = path+"dna.fasta"
    analyze_fasta(fasta_file)

Longest sequence: 142022655 (2646 bp)
Shortest sequence: 142022655 (2646 bp)

Gene 142022655: Longest ORF in Frame 3
Start: 635, End: 2456, Length: 1821
ORF sequence (first 60 bp): ATGAACAGCGGGGCGAGCAAGCCGCCGGCCGTCACGGGGTCCATCACGAGGGACAGCAGC...



The above result considers only one sequence because we are parsing it through the gene identifiers (2nd column) which is "142022655" same for all the sequences.


Below I have considered seq.id from the seqIO library which takes the combination of columns (e.g. "gi|142022655|gb|EQ086233.1|91") as an unique identifier for the sequence. Hence the result that I get below after considering this are:

Longest sequence: gi|142022655|gb|EQ086233.1|255 (4894 bp)
Shortest sequence: gi|142022655|gb|EQ086233.1|346 (115 bp)

gi|142022655|gb|EQ086233.1|91: Longest ORF in Frame 1
Start: 978, End: 2274, Length: 1296
ORF sequence (first 60 bp): ATGCATCATCCCGACGCGCAACGCCAGCTGGTTGCGGCCCGACGACTGCCCGGCCGTGCC...

gi|142022655|gb|EQ086233.1|304: Longest ORF in Frame 3
Start: 620, End: 767, Length: 147
ORF sequence (first 60 bp): ATGATGATCTTGTTCTTCGGTGTCGTCGTGGACAGCGACGCGATCTGCTGTCCGTTCAGC...

gi|142022655|gb|EQ086233.1|255: Longest ORF in Frame 1
Start: 291, End: 1734, Length: 1443
ORF sequence (first 60 bp): ATGACCATGTCCCGCGGCCCGCCGACGCGACGCCACCCCGTGCCGCGCCGTGTCATGTGC...

gi|142022655|gb|EQ086233.1|45: Longest ORF in Frame 1
Start: 384, End: 2778, Length: 2394
ORF sequence (first 60 bp): ATGGAGAAACAGTCTCGCGTTACGCGCGACGGTCGCGGGAGAGTTCTATGCGGTCATCGC...

gi|142022655|gb|EQ086233.1|396: Longest ORF in Frame 2
Start: 1516, End: 2797, Length: 1281
ORF sequence (first 60 bp): ATGCCGCCCGCCTGTGGGCGCGCGGCCCGCAAGGGCGGCTGGCCATGCAGGCCAGCGCCG...

gi|142022655|gb|EQ086233.1|250: Longest ORF in Frame 1
Start: 561, End: 2121, Length: 1560
ORF sequence (first 60 bp): ATGCCGGCGCTGGTGGTTGAACAGGTCGGCCGAGCCCGCAAGGCGCGCGGCGAGCTCCGG...

gi|142022655|gb|EQ086233.1|322: Longest ORF in Frame 3
Start: 89, End: 278, Length: 189
ORF sequence (first 60 bp): ATGCGTACGACGCAGCCGTGCTGCAGTACAAGGCGACGGTGCTGTCCGCATTCGGCGACG...

gi|142022655|gb|EQ086233.1|88: Longest ORF in Frame 2
Start: 181, End: 319, Length: 138
ORF sequence (first 60 bp): ATGGGTCGTGGTCGACGTGAAGAACGCCGGGCTCGACCAGCCGTGCACGGCGCTGAACTG...

gi|142022655|gb|EQ086233.1|594: Longest ORF in Frame 3
Start: 65, End: 278, Length: 213
ORF sequence (first 60 bp): ATGATCGGCGAAATCGACATCTTCGGCGTATTCGTGCCGGCCCCGCTCGTGCTGATGCTG...

gi|142022655|gb|EQ086233.1|293: Longest ORF in Frame 3
Start: 2333, End: 3044, Length: 711
ORF sequence (first 60 bp): ATGCCGGTTGCGCTGATCCCAGTTGCCGAACAGCGATTCGAATGCAACCGCGGCGACCGC...

gi|142022655|gb|EQ086233.1|75: Longest ORF in Frame 2
Start: 256, End: 760, Length: 504
ORF sequence (first 60 bp): ATGATTGCGCGGATCGGAGCTGATCGACACCGGGATGCCGAGCCGCGACTGCTCCGCCAG...

gi|142022655|gb|EQ086233.1|454: Longest ORF in Frame 3
Start: 3095, End: 4496, Length: 1401
ORF sequence (first 60 bp): ATGGCCGATCAGTTCGCGGATCTCCTTCGCCGCCGACGCGCAGCGCTGCGCGAGGCCGCG...

gi|142022655|gb|EQ086233.1|16: Longest ORF in Frame 3
Start: 1439, End: 3083, Length: 1644
ORF sequence (first 60 bp): ATGCGGGCCATCCTGCATCGCCGCCTTTCGTTCCACCCGGGCCGGCATCGAGTGATGCCG...

gi|142022655|gb|EQ086233.1|584: Longest ORF in Frame 3
Start: 347, End: 479, Length: 132
ORF sequence (first 60 bp): ATGCTCCAGACTTCCAGGCAAAATGGTGCAATGTGCGTACCACGCTTGACCAACTCGTCA...

gi|142022655|gb|EQ086233.1|4: Longest ORF in Frame 1
Start: 444, End: 693, Length: 249
ORF sequence (first 60 bp): ATGGGTGGCCAGCAGAGGCCGGGAAACCCAGGCACTCGCAGACACGATCGCCATCGGCGA...

gi|142022655|gb|EQ086233.1|277: Longest ORF in Frame 2
Start: 679, End: 958, Length: 279
ORF sequence (first 60 bp): ATGGGCGAACGGGCATTTCCGCGACTATGCGCTCGCGCAGAACAGCGAGACGCGCGTGCC...

gi|142022655|gb|EQ086233.1|346: No valid ORF found

gi|142022655|gb|EQ086233.1|527: Longest ORF in Frame 3
Start: 635, End: 2456, Length: 1821
ORF sequence (first 60 bp): ATGAACAGCGGGGCGAGCAAGCCGCCGGCCGTCACGGGGTCCATCACGAGGGACAGCAGC...


In [12]:
# Finds longest/shortest DNA sequence and longest ORFs in each frame


# Define codons
START_CODON = "ATG"
STOP_CODONS = {"TAA", "TAG", "TGA"}

#Find all ORFs in all three forward reading frames.
def find_orfs(seq):
    seq = str(seq).upper()
    frames = {1: [], 2: [], 3: []}

    # iterate through each frame
    for frame in range(3):
        start_positions = []
        for i in range(frame, len(seq)-2, 3):
            codon = seq[i:i+3]
            if codon == START_CODON:
                start_positions.append(i)
            elif codon in STOP_CODONS:
                # For each open start, close the ORF
                for start in start_positions:
                    orf_seq = seq[start:i+3]
                    frames[frame+1].append((start, i+3, orf_seq, len(orf_seq)))
                start_positions = []  # reset after stop
    return frames

# Return the longest ORF among all frames.
def find_longest_orf(frames):
    longest = None
    for frame, orfs in frames.items():
        for orf in orfs:
            if longest is None or orf[3] > longest[1][3]:
                longest = (frame, orf)
    return longest  # (frame_number, (start, end, orf_seq, length))

# Main function to analyze DNA sequences.
def analyze_fasta(fasta_file):
    records = list(SeqIO.parse(fasta_file, "fasta"))

    # Find longest and shortest sequences
    longest_record = max(records, key=lambda r: len(r.seq))
    shortest_record = min(records, key=lambda r: len(r.seq))

    print(f"Longest sequence: {longest_record.id} ({len(longest_record.seq)} bp)")
    print(f"Shortest sequence: {shortest_record.id} ({len(shortest_record.seq)} bp)\n")

    # Find longest ORFs
    for record in records:
        frames = find_orfs(record.seq)
        longest_orf = find_longest_orf(frames)

        if longest_orf:
            frame_num, (start, end, orf_seq, length) = longest_orf
            print(f"{record.id}: Longest ORF in Frame {frame_num}")
            print(f"Start: {start}, End: {end}, Length: {length}")
            print(f"ORF sequence (first 60 bp): {orf_seq[:60]}...\n")
        else:
            print(f"{record.id}: No valid ORF found\n")


if __name__ == "__main__":
    fasta_file = path+"dna.fasta"
    analyze_fasta(fasta_file)

Longest sequence: gi|142022655|gb|EQ086233.1|255 (4894 bp)
Shortest sequence: gi|142022655|gb|EQ086233.1|346 (115 bp)

gi|142022655|gb|EQ086233.1|91: Longest ORF in Frame 1
Start: 978, End: 2274, Length: 1296
ORF sequence (first 60 bp): ATGCATCATCCCGACGCGCAACGCCAGCTGGTTGCGGCCCGACGACTGCCCGGCCGTGCC...

gi|142022655|gb|EQ086233.1|304: Longest ORF in Frame 3
Start: 620, End: 767, Length: 147
ORF sequence (first 60 bp): ATGATGATCTTGTTCTTCGGTGTCGTCGTGGACAGCGACGCGATCTGCTGTCCGTTCAGC...

gi|142022655|gb|EQ086233.1|255: Longest ORF in Frame 1
Start: 291, End: 1734, Length: 1443
ORF sequence (first 60 bp): ATGACCATGTCCCGCGGCCCGCCGACGCGACGCCACCCCGTGCCGCGCCGTGTCATGTGC...

gi|142022655|gb|EQ086233.1|45: Longest ORF in Frame 1
Start: 384, End: 2778, Length: 2394
ORF sequence (first 60 bp): ATGGAGAAACAGTCTCGCGTTACGCGCGACGGTCGCGGGAGAGTTCTATGCGGTCATCGC...

gi|142022655|gb|EQ086233.1|396: Longest ORF in Frame 2
Start: 1516, End: 2797, Length: 1281
ORF sequence (first 60 bp): ATGCCGCCCGCCTGTGGGCGCGCGGCCCG